File: sdks/java/src/main/java/ai/beatbox/model/Source.java
package ai.beatbox.model;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

/**
 * The program to run, a tagged union on the {@code kind} discriminator.
 *
 * <p>Use the static factory methods for the common cases, e.g.
 * {@code Source.wasmWat("(module ...)")}. The remote API only accepts {@code wasm_wat} and
 * {@code wasm_bytes_base64} for the {@code wasm} lane.
 */
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "kind")
@JsonSubTypes({
        @JsonSubTypes.Type(value = Source.Inline.class, name = "inline"),
        @JsonSubTypes.Type(value = Source.WasmFile.class, name = "wasm_file"),
        @JsonSubTypes.Type(value = Source.WasmWat.class, name = "wasm_wat"),
        @JsonSubTypes.Type(value = Source.WasmBytesBase64.class, name = "wasm_bytes_base64"),
        @JsonSubTypes.Type(value = Source.ModuleRef.class, name = "module_ref"),
})
public sealed interface Source
        permits Source.Inline, Source.WasmFile, Source.WasmWat, Source.WasmBytesBase64, Source.ModuleRef {

    /** Inline source code interpreted by the selected lane. */
    record Inline(String code) implements Source {
    }

    /** A wasm module already present on the daemon host at {@code path}. */
    record WasmFile(String path) implements Source {
    }

    /** WebAssembly text format, compiled by the daemon. */
    record WasmWat(String text) implements Source {
    }

    /** A base64-encoded wasm binary. */
    record WasmBytesBase64(String bytes) implements Source {
    }

    /** A reference to a previously uploaded module by its sha256 digest. */
    record ModuleRef(String sha256) implements Source {
    }

    static Source inline(String code) {
        return new Inline(code);
    }

    static Source wasmFile(String path) {
        return new WasmFile(path);
    }

    static Source wasmWat(String text) {
        return new WasmWat(text);
    }

    static Source wasmBytesBase64(String bytes) {
        return new WasmBytesBase64(bytes);
    }

    static Source moduleRef(String sha256) {
        return new ModuleRef(sha256);
    }
}

File: sdks/csharp/src/Beatbox/Source.cs
using System.Text.Json.Serialization;

namespace Beatbox;

/// <summary>
/// Program source for an execution. A tagged union discriminated by
/// <see cref="Kind"/>; construct one with the factory methods rather than setting
/// fields directly. Only the field relevant to the chosen kind is serialized.
/// </summary>
public sealed record Source
{
    /// <summary>
    /// Discriminator: one of <c>inline</c>, <c>wasm_file</c>, <c>wasm_wat</c>,
    /// <c>wasm_bytes_base64</c>, or <c>module_ref</c>.
    /// </summary>
    [JsonPropertyName("kind")]
    public string Kind { get; init; } = "";

    /// <summary>Inline source code (kind <c>inline</c>).</summary>
    [JsonPropertyName("code")]
    public string? Code { get; init; }

    /// <summary>Path to a wasm module on the host (kind <c>wasm_file</c>).</summary>
    [JsonPropertyName("path")]
    public string? Path { get; init; }

    /// <summary>WebAssembly text (kind <c>wasm_wat</c>).</summary>
    [JsonPropertyName("text")]
    public string? Text { get; init; }

    /// <summary>Base64-encoded wasm module bytes (kind <c>wasm_bytes_base64</c>).</summary>
    [JsonPropertyName("bytes")]
    public string? Bytes { get; init; }

    /// <summary>Content-addressed module reference (kind <c>module_ref</c>).</summary>
    [JsonPropertyName("sha256")]
    public string? Sha256 { get; init; }

    /// <summary>Inline source code, interpreted by the target lane.</summary>
    public static Source Inline(string code) => new() { Kind = "inline", Code = code };

    /// <summary>A wasm module already present on the host filesystem.</summary>
    public static Source WasmFile(string path) => new() { Kind = "wasm_file", Path = path };

    /// <summary>WebAssembly text (WAT) compiled by the daemon.</summary>
    public static Source WasmWat(string text) => new() { Kind = "wasm_wat", Text = text };

    /// <summary>A base64-encoded wasm binary.</summary>
    public static Source WasmBytesBase64(string bytes) => new() { Kind = "wasm_bytes_base64", Bytes = bytes };

    /// <summary>A previously uploaded module referenced by its sha256 digest.</summary>
    public static Source ModuleRef(string sha256) => new() { Kind = "module_ref", Sha256 = sha256 };
}
